### PR TITLE
layer file download: final rename: fix durability

### PR DESCRIPTION
**Pull request by problame**
Before this PR, the layer file download code would fsync the inode after
rename instead of the timeline directory. That is not in line with what
a comment further up says we're doing, and it's obviously not achieving
the goal of making the rename durable.

part of https://github.com/neondatabase/neon/issues/6663

### CODE REVIEW

File: pageserver/src/tenant/remote_timeline_client/download.rs
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client/download.rs:160]: Hm, I was going to argue that we will this PR after #6378 , so, the VirtualFile::sync_all() will be asynchronous on tokio-epoll-uring. 

But, with `std-fs`, it will block the executor thread.
So, would need to bifurcate this.

File: pageserver/src/tenant/remote_timeline_client/download.rs
Comment (koivunej) [inline pageserver/src/tenant/remote_timeline_client/download.rs:150]: The comment also mentions that we should fsync the new name with `fsync(new)`. Perhaps the comment is wrong in this regard as the name is stored on the directory entry AFAIR?

Comment (koivunej) [reply pageserver/src/tenant/remote_timeline_client/download.rs:150]: `fsync(new)` might be some ext3 with strange mode requirement, which we don't necessarily need to support. Could at least comment here the discrepency?

**Review by koivunej — APPROVED** [review]
Regardless of `fsync(new)` after rename or the discrepency due to lack of it to the comment above, this is already a good fix. Good note on the sync_all blocking on std-fs, bifurcation.

File: pageserver/src/tenant/remote_timeline_client/download.rs
Comment (problame) [reply pageserver/src/tenant/remote_timeline_client/download.rs]: I pushed https://github.com/neondatabase/neon/pull/6991/commits/ca616ee10bfd95104fff51f0a195966153e8b9ff to add the bifurcation.

Yet, before #6378  is merged, which makes `sync_all()` truly async for tokio-epoll-uring, we shouldn't merge this PR.

File: pageserver/src/tenant/remote_timeline_client/download.rs
Comment (problame) [reply pageserver/src/tenant/remote_timeline_client/download.rs:150]: > Perhaps the comment is wrong in this regard as the name is stored on the directory entry AFAIR?

The comment is wrong IMO, the second `fsync(new)` is not necessary if the first fsync is a true fsync and not an `fdatasync`.
I think Postgres's `file_utils.c` is also too conservative.

But, let's not have this discussion here.

I'm inclined to leave the comment unchanged and will revisit this at some later point / talk with postgres people why they think it's necessary.

Anyway, the rest of the pageserver code base also doesn't do the second `fsync()` whenever it needs to make created files durable, so, let's be consistent about it.

**Review by koivunej — APPROVED** [review]

**Pull request merged**
